[RFC 0068] Minimal daemon (#68)
NixOS/rfcs: rfcs/0000-minimal-daemon.md

---
feature: minimal-daemon
start-date: 2020-5-28
author: John Ericson
co-authors: (find a buddy later to help out with the RFC)
shepherd-team: (names, to be nominated and accepted by RFC steering committee)
shepherd-leader: (name to be appointed by RFC steering committee)
related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

`nix-daemon` should be a separate executable.

# Motivation
[motivation]: #motivation

With flakes and other development, we are moving towards a more "batteries included" Nix command line.
We don't want any of those features in the daemon, however, because the daemon is a special trusted process that we should strive to keep as simple as possible.

> \[This is comparable to a compilation pipeline, with a concise intermediate representation nicer user-facing features "desugar" into.\]

Reviewer: Nit that I think slightly improves readability:

```suggestion
\[This is comparable to a compilation pipeline, with a concise intermediate representation that nicer user-facing features "desugar" into.\]
```

Or maybe it could be reworded as something like: 
```suggestion
\[This is comparable to a compilation pipeline, where nicer user-facing features "desugar" into a concise intermediate representation.\]
```

There are many things we could do about this, but I mainly want to establish some rough consensus around the problem while taking a small step to signal that consensus.

> Originally, each Nix command was its own executable, but then we combined them into one executable.

Reviewer: ```suggestion
Originally, each Nix command was its own executable, which we then combined them into one executable.
```

Maybe?

I think this is fine for the main user-facing commands, but not good for the daemon.

\[This RFC is completely separate from the ongoing IPFS work.]

# Detailed design
[design]: #detailed-design

- `nix-daemon` will be a separate executable that only links the nix libraries it needs.

>   \[At this time, those libraries are `libnixutil`, `libnixutil`, and `libnixrust`, but this is subject to change.\]

Reviewer: Should there be another library listed instead of two `libnixutil`s? (Being unfamiliar with the Nix tool itself, I can't suggest what that might be.)

Reviewer 2: `libnixstore` I guess

Author: ```suggestion
  \[At this time, those libraries are `libnixutil`, `libnixstore`, and `libnixrust`, but this is subject to change.\]
```

> - `nix-daemon` should never need to understand the expression language and depend  `libnixexpr`.

Reviewer 2: :+1: since the separate nix-daemon is also what allows things like guix

Reviewer 3: What do you mean _it allows guix_?

Reviewer 4: In the sense the daemon can work in other frameworks not directly related to Nix language. Guix can use the same daemon as us and they don't need to read a single line of Nix code, only Scheme (and maybe C++).

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

I certainly hope there are no interactions!
One of the bad things we should seek to prevent with this is the daemon unintentionally growing dependencies on more of the code base.

# Drawbacks
[drawbacks]: #drawbacks

Not much.
Build rules perhaps are slightly more complex as there are both separate and independent executables.

# Alternatives
[alternatives]: #alternatives

 - Do nothing.

 - Something more invasive.
   But I much rather save that for later.

# Unresolved questions
[unresolved]: #unresolved-questions

No known unknowns.

# Future work
[future]: #future-work

> There's lots of future work we could do in the vain of modularity.

Reviewer: ```suggestion
There's lots of future work we could do in the vein of modularity.
```

---
feature: minimal-daemon
start-date: 2020-5-28
author: John Ericson
co-authors: (find a buddy later to help out with the RFC)

> shepherd-team: (names, to be nominated and accepted by RFC steering committee)

Reviewer 5: Please add @edolstra, @regnat and @Kloenk to the shepherd team.

> shepherd-leader: (name to be appointed by RFC steering committee)

Reviewer 5: Please add @edolstra as a leader.

- `nix-daemon` should never need to understand the expression language and depend  `libnixexpr`.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

I certainly hope there are no interactions!
One of the bad things we should seek to prevent with this is the daemon unintentionally growing dependencies on more of the code base.

# Drawbacks
[drawbacks]: #drawbacks

Installation is slightly bigger as the two binaries (`nix` and `nix-daemon`) have some redundancy.
Build rules perhaps are slightly more complex as there are both separate and independent executables.

# Alternatives
[alternatives]: #alternatives

 - Do nothing.

 - Something more invasive, such as packaging the libraries and commands separately, or intending the libaries for widespread public consumption.
   But I much rather save that for later, as such steps would be far more controversial.

# Unresolved questions
[unresolved]: #unresolved-questions

No known unknowns.

# Future work
[future]: #future-work

I think most people for this will have future plans they wish to persue in the name of modularity.

> But I don't expect everyone to agree on waht exactl those plans should be.

Reviewer 6: spelling
```suggestion
But I don't expect everyone to agree on what exactly those plans should be.
```

---
feature: minimal-daemon
start-date: 2020-5-28
author: John Ericson
co-authors: (find a buddy later to help out with the RFC)
shepherd-team: (names, to be nominated and accepted by RFC steering committee)
shepherd-leader: (name to be appointed by RFC steering committee)
related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

`nix-daemon` should be a separate executable.

# Motivation
[motivation]: #motivation

With flakes and other development, we are moving towards a more "batteries included" Nix command line.
We don't want any of those features in the daemon, however, because the daemon is a special trusted process that we should strive to keep as simple as possible.

> \[This is comparable to a compilation pipeline, with a concise intermediate representation that nicer user-facing features "desugar" into.\]

Reviewer 7: I would either remove the square brackets, or remove this.

There are many things we could do about this, but I mainly want to establish some rough consensus around the problem while taking a small step to signal that consensus.
Originally, each Nix command was its own executable, but then we combined them into one executable.
I think this is fine for the main user-facing commands, but not good for the daemon.

> Finally, it's probably best not to give the daemon---a long lived process running with elevated privileges---access to tons of dead code.

Reviewer 7: Can we specify exactly what privileges the daemon is running with, here? I think it's pertinent to the motivations of this RFC.

Reviewer 7: And what is meant here by dead code?

Author: Will explain the "dead code". The elevated privileges thing I am not sure what to right: in theory the nix daemon just needs to be make sandboxes and access the store, but in practice we run it as root last I checked.

Reviewer 8: Yes, it is run as Root. Currently the source doesn't allow anything else. We could check for CAPs, but we only check for Root prior to remounting the store in rw

C++ doesn't exactly prevent memory errors, and that dead code is just more fodder to be used in some low-level attack.

> There are other solutions to this in the long term, but this is the easiest solution in the short term.

Reviewer 7: Would be good to mention the other solutions in *Future Work*, methinks.

> \[This RFC is completely separate from the ongoing IPFS work.]

Reviewer 7: Don't think this should stay in the final version of the RFC. Not sure if that was your intent :)

Author: Fair. I announced that around the same time I opened this, but in hindsight I don't think it is very confusing.

# Detailed design
[design]: #detailed-design

- `nix-daemon` will be a separate executable that only links the nix libraries it needs.
  \[At this time, those libraries are `libnixutil`, `libnixstore`, and `libnixrust`, but this is subject to change.\]

- `nix-daemon` should never need to understand the expression language and depend  `libnixexpr`.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

I certainly hope there are no interactions!

> One of the bad things we should seek to prevent with this is the daemon unintentionally growing dependencies on more of the code base.

Reviewer 7: I think this could be moved to the *Motivations* section.

# Drawbacks
[drawbacks]: #drawbacks

> Not much.

Reviewer 5: Maybe performance.

Author: ```suggestion
Installation is slightly bigger as the two binaries (`nix` and `nix-daemon`) have some redundancy.
```

Author: If anything the performance of the daemon could be better, but yes it does add complexity (which I confusingly did remark about below) and it also increase space usage, which I added here in place of "not much".

> Build rules perhaps are slightly more complex as there are both separate and independent executables.

Reviewer 7: What would it take to remove the *perhaps* and know for sure? :)

# Alternatives
[alternatives]: #alternatives

 - Do nothing.

 - Something more invasive.

>    But I much rather save that for later.

Reviewer 7: I would remove this, as it's a personal opinion that doesn't provide much information. Or maybe explain what that something more invasive is, and why it should be postponed for later? :)

Author: Will make it more concrete.

# Unresolved questions
[unresolved]: #unresolved-questions

No known unknowns.

# Future work
[future]: #future-work

> There's lots of future work we could do in the vein of modularity.

Reviewer 7: Same here, I would try to be (a bit more) specific.

Author: Well, I'm trying to elude to various things people have told me they'd like to see happen over the years, without poisoning the specifics. I rather have this stuff organically come back in the RFC shepherd meetings.

Originally, each Nix command was its own executable, but then we combined them into one executable.
I think this is fine for the main user-facing commands, but not good for the daemon.

Finally, it's probably best not to give the daemon---a long lived process running with elevated privileges---access to tons of dead code.

> C++ doesn't exactly prevent memory errors, and that dead code just more fodder to be used in some low-level attack.

Reviewer: ```suggestion
C++ doesn't exactly prevent memory errors, and that dead code is just more fodder to be used in some low-level attack.
```
